Explicit Dynamics: Initial Condition for Solution Time Derivatives

ddm42 opened the pull request:
- Added capability of using ICs to affect old states

closes #28073

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2317920/) on 78bd3c8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28083/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28083/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6349167f4073a10e927fc73d403dd65fd370db6c`

dschwen commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -2318,6 +2318,9 @@ class FEProblemBase : public SubProblem, public Restartable
   virtual void setCurrentBoundaryID(BoundaryID bid, const THREAD_ID tid) override;
 
+  // Used for setting previous states
+  int global_current_state;
```
```suggestion
  int _global_current_state;
```
we prefix member variables withe an underscore

dschwen commented on framework/src/ics/InitialConditionBase.C:
```diff
@@ -28,9 +30,12 @@ InitialConditionBase::validParams()
-
   params.addParamNamesToGroup("ignore_uo_dependency", "Advanced");
 
+  params.addParam<int>("state", 0, "Specify which state being set.");
```
```suggestion
  params.addParam<int>("state", 0, "Specify which state is being set, with 0 being the current state, 1 the old state, 2 the older state, etc..");
```

dschwen commented on framework/src/ics/InitialConditionBase.C-e:
```diff
@@ -0,0 +1,100 @@
+//* This file is part of the MOOSE framework
```
What is this file? `framework/src/ics/InitialConditionBase.C-e`

dschwen commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -43,43 +45,45 @@ InitialConditionWarehouse::addObject(std::shared_ptr<InitialConditionBase> objec
   {
-    std::map<std::string, std::set<SubdomainID>>::const_iterator iter =
-        _block_ics[tid].find(var.name());
+    std::map<std::tuple<std::string, int>, std::set<SubdomainID>>::const_iterator iter =
```
```suggestion
    auto iter =
```

dschwen replied:
I encourage more use of `auto`

dschwen commented on framework/src/ics/InitialConditionWarehouse.C-e:
```diff
@@ -0,0 +1,104 @@
+//* This file is part of the MOOSE framework
```
remove this from the PR, too

ddm42 replied:
will do. Something was wrong with my clang-format and its interaction with Mac.

dschwen commented on framework/src/loops/ComputeInitialConditionThread.C:
```diff
@@ -93,8 +93,26 @@ ComputeInitialConditionThread::operator()(const ConstElemRange & range)
+      int loop_state;
+      for (loop_state = _fe_problem.global_current_state; loop_state >= 0; loop_state--)
+      {
+        for (auto ic : vec)
+        {
+          if (ic->_my_state == loop_state)
+          {
+            ic->compute();
+            break_loop = true;
+          }
+        }
+        if (break_loop == true)
+        {
+          break;
+        }
```
```suggestion
        for (auto ic : vec)
          if (ic->_my_state == loop_state)
          {
            ic->compute();
            break_loop = true;
          }

        if (break_loop == true)
          break;
```

I'm not sure i understand the `break_loop` logic. Wouldn't that skip states for all ICs if one of them affects an older state? Even if they act on different variables?

ddm42 replied:
This for-loop logic was implemented inside an existing loop that runs for each relevant variable. So the ics being considered are all for the same variable. Does that make sense? check lines 78-81.

ddm42 reviewed (commented):
Thanks for the comments. I'll commit and test them

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2327988/) on 154a861 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28083/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28083/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0797b81f7e4481a0612d672fe053e18bc97e185a`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2328402/) on 17cb247 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28083/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28083/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0797b81f7e4481a0612d672fe053e18bc97e185a`

dschwen commented on framework/src/ics/InitialConditionBase.C:
```diff
@@ -28,9 +30,15 @@ InitialConditionBase::validParams()
-
   params.addParamNamesToGroup("ignore_uo_dependency", "Advanced");
 
+  params.addParam<int>("state",
+                       0,
+                       "Specify which state is being set, with 0 being the current state, 1 the "
+                       "old state, 2 the older state, etc..");
```
Let's try this:

```suggestion
  MooseEnum stateEnum("CURRENT=0 OLD=1 OLDER=3", "CURRENT", /* allow_out_of_range = */ true);
  params.addParam<int>("state", stateEnum, "Specify state that is being set.");
```

dschwen replied:
Let's just set `allow_out_of_range` to false and not support states older than `OLDER` for now. (so, simply remove the `, /* allow_out_of_range = */ true` above

dschwen commented on framework/src/ics/InitialConditionBase.C:
```diff
@@ -52,6 +60,7 @@ InitialConditionBase::InitialConditionBase(const InputParameters & parameters)
     DependencyResolverInterface(),
     Restartable(this, "InitialConditionBases"),
     ElementIDInterface(this),
+    _my_state(getParam<int>("state")),
```
```suggestion
    _my_state(getParam<MooseEnum>("state")),
```

The enum should automatically cast to int

dschwen commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -43,43 +45,44 @@ InitialConditionWarehouse::addObject(std::shared_ptr<InitialConditionBase> objec
 
-    std::map<std::string, std::set<BoundaryID>>::const_iterator iter =
-        _boundary_ics[tid].find(var.name());
+    std::map<std::tuple<std::string, int>, std::set<BoundaryID>>::const_iterator iter =
```
```suggestion
    auto iter =
```

let's use `auto` where we can

dschwen commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -43,43 +45,44 @@ InitialConditionWarehouse::addObject(std::shared_ptr<InitialConditionBase> objec
   {
-    std::map<std::string, std::set<SubdomainID>>::const_iterator iter =
-        _block_ics[tid].find(var.name());
+    std::map<std::tuple<std::string, int>, std::set<SubdomainID>>::const_iterator iter =
```
same her

dschwen commented on python/MooseDocs/test/gold/latex/extensions/include.tex:
```diff
@@ -1,9 +1,9 @@
 \begin{quote}
-\par Congress shall make no law respecting an establishment of religion, or prohibiting the free exercise thereof; or abridging the freedom of speech, or of the press; or the right of the people peaceably to assemble, and to petition the Government for a redress of grievances. 
-\par A well regulated Militia, being necessary to the security of a free State, the right of the people to keep and bear Arms, shall not be infringed. 
+\par Congress shall make no law respecting an establishment of religion, or prohibiting the free exercise thereof; or abridging the freedom of speech, or of the press; or the right of the people peaceably to assemble, and to petition the Government for a redress of grievances.
```
You have a bunch of TeX files that just have whitespace changes. Those shoudl be removed from this PR. I can help you with that.

ddm42 committed:
Explicit Dynamics: Initial Condition for Solution Time Derivatives

- added capability to affect old states using initial conditions
- added tests for old state ICs (solid_mechanics/test/tests/old_state_ic)
- added capability to copy old states of variables in CopyValueAux
- edited CopyValueAux tests

fixes #28073

dschwen commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -35,6 +36,7 @@ InitialConditionWarehouse::addObject(std::shared_ptr<InitialConditionBase> objec
 {
   // Check that when object is boundary restricted that the variable is nodal
   const MooseVariableFEBase & var = object->variable();
+  const std::tuple<std::string, int> ic_key = std::tuple(var.name(), object->_my_state);
```
```suggestion
  const auto ic_key = std::tuple(var.name(), object->_my_state);
```

hugary1995 replied:
hmm I see this conversation is marked as resolved, but the suggested change hasn't been applied. Is github broken?

ddm42 replied:
No I was marking them resolved because I made the change on my end. Just trying to keep track of the ones I have left.

dschwen commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -43,43 +45,42 @@ InitialConditionWarehouse::addObject(std::shared_ptr<InitialConditionBase> objec
 
-    std::map<std::string, std::set<BoundaryID>>::const_iterator iter =
-        _boundary_ics[tid].find(var.name());
+    auto iter = _boundary_ics[tid].find(ic_key);
```
```suggestion
    const auto iter = _boundary_ics[tid].find(ic_key);
```

Sorry, that should have been const

dschwen commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -43,43 +45,42 @@ InitialConditionWarehouse::addObject(std::shared_ptr<InitialConditionBase> objec
   {
-    std::map<std::string, std::set<SubdomainID>>::const_iterator iter =
-        _block_ics[tid].find(var.name());
+    auto iter = _block_ics[tid].find(ic_key);
```
same

ddm42 replied:
What do you mean by same? the type is already "auto"?

dschwen reviewed (commented):
Looks fine to me if those minor comments are addressed.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2346601/) on 3bada02 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28083/site)

This comment will be updated on new commits.

hugary1995 commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -2348,6 +2348,9 @@ class FEProblemBase : public SubProblem, public Restartable
   virtual void setCurrentBoundaryID(BoundaryID bid, const THREAD_ID tid) override;
 
+  // Used for setting previous states
+  int _global_current_state;
```
Is this public? It's kinda difficult to check in the web browser, but if so move it to protected at least.

hugary1995 commented on framework/include/timeintegrators/CentralDifference.h:
```diff
@@ -10,6 +10,7 @@
 #pragma once
 
 #include "ActuallyExplicitEuler.h"
+#include "libmesh/numeric_vector.h"
```
why is this?

hugary1995 commented on framework/src/ics/InitialConditionBase.C:
```diff
@@ -8,9 +8,11 @@
 #include "SystemBase.h"
 #include "MooseVariableFE.h"
 #include "UserObject.h"
+#include "libmesh/libmesh_common.h"
```
why this?

ddm42 replied:
I'll remove this. It must have added it automatically.

hugary1995 commented on framework/src/ics/InitialConditionBase.C:
```diff
@@ -28,9 +30,13 @@ InitialConditionBase::validParams()
   params.addParamNamesToGroup("ignore_uo_dependency", "Advanced");
 
+  MooseEnum stateEnum("CURRENT=0 OLD=1 OLDER=2", "CURRENT");
+  params.addParam<MooseEnum>("state", stateEnum, "Specify state that is being set.");
```
Can you expand the description a bit more?

hugary1995 commented on framework/src/ics/InitialConditionBase.C:
```diff
@@ -28,9 +30,13 @@ InitialConditionBase::validParams()
+  MooseEnum stateEnum("CURRENT=0 OLD=1 OLDER=2", "CURRENT");
+  params.addParam<MooseEnum>("state", stateEnum, "Specify state that is being set.");
+
+  params.addParamNamesToGroup("state", "Advanced");
```
I feel like this doesn't have to go into the advanced group, but this is apparently debatable.

hugary1995 commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -43,43 +45,42 @@ InitialConditionWarehouse::addObject(std::shared_ptr<InitialConditionBase> objec
       mooseError("The initial condition '",
                  object->name(),
-                 "' is being defined on a boundary that already has an initial condition defined.");
+                 "' is being defined on a boundary that already has an initial condition defined "
+                 "(for the variable and state combo).");
```
This is user-facing so let's make the wording a bit more professional. For "variable and state combo", also tell people which combo it is.

hugary1995 commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -43,43 +45,42 @@ InitialConditionWarehouse::addObject(std::shared_ptr<InitialConditionBase> objec
       mooseError("The initial condition '",
                  object->name(),
-                 "' is being defined on a block that already has an initial condition defined.");
+                 "' is being defined on a block that already has an initial condition defined (for "
+                 "the variable and state combo).");
```
Same suggestion as above.

hugary1995 commented on framework/src/ics/InitialConditionWarehouse.C:
```diff
@@ -43,43 +45,42 @@ InitialConditionWarehouse::addObject(std::shared_ptr<InitialConditionBase> objec
       mooseError("The initial condition '",
                  object->name(),
-                 "' is being defined on a block that already has an initial condition defined.");
+                 "' is being defined on a block that already has an initial condition defined (for "
+                 "the variable and state combo).");
```
Same suggestion on wording.

hugary1995 commented on framework/include/ics/InitialConditionBase.h:
```diff
@@ -91,6 +91,9 @@ class InitialConditionBase : public MooseObject,
   virtual const std::set<std::string> & getSuppliedItems() override;
 
+  // variable used when applying initial conditions to previous states
+  int _my_state;
```
Is this public? Move it to protected at least.

hugary1995 commented on framework/src/loops/ComputeInitialConditionThread.C:
```diff
@@ -93,8 +93,23 @@ ComputeInitialConditionThread::operator()(const ConstElemRange & range)
+      /* Logic for running only old state initial conditions if they exist, but
+          running ICs normally if the old state ICs don't exist.
+      */
+      bool break_loop = false;
+      int loop_state;
+      for (loop_state = _fe_problem._global_current_state; loop_state >= 0; loop_state--)
+      {
+        for (auto ic : vec)
+          if (ic->_my_state == loop_state)
+          {
+            ic->compute();
+            break_loop = true;
+          }
+
+        if (break_loop == true)
+          break;
+      }
```
You could help me understand this better... Why do we immediately break after one IC is computed? What if I have multiple IC objects for the same variable with the same state, just defined on different blocks/boundaries, does this mean only one of them gets computed?

As a side note, directly accessing members with an underscore prefix is not the best practice in MOOSE.

ddm42 replied:
We are not breaking after one IC is computed. Lines 104-108 compute for every ic in the vector before it has a chance to break. 

However, I am open to suggestions for better logic. Just remember that vec is a vector of ic's for only ONE variable. So other variables will not be affected.

ddm42 replied:
I used a member variable _fe_problem._global_current_state because we needed to transfer the loop state from FEProblemBase::projectSolution(). Is there another location for a loop variable that would be available in both files? That was why I chose to implement it this way. But I am open to suggestions.

hugary1995 replied:
I need to look at this more closely. I suggest other reviewers should also double check this.

ddm42 replied:
I would love to implement it a cleaner way. Thanks for checking it out.

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3332,64 +3335,86 @@ FEProblemBase::projectSolution()
-  Threads::parallel_reduce(elem_range, cic);
 
-  if (haveFV())
+  // loop over _global_current_state = 2, 1, 0
+  for (_global_current_state = 1; _global_current_state >= 0; _global_current_state--)
```
But this is not 2, 1, 0 -- this is 1, 0. Now I wonder if you have test coverage for checking the older IC being set correctly...

ddm42 replied:
I don't. I shall add some.

hugary1995 commented on framework/src/timeintegrators/CentralDifference.C:
```diff
@@ -114,3 +114,30 @@ CentralDifference::computeTimeDerivatives()
     u_dot_factor.close();
   }
 }
+
+template <typename T, typename T2, typename T3, typename T4>
+void
+CentralDifference::computeTimeDerivativeHelper(T & u_dot,
+                                               T2 & u_dotdot,
+                                               const T3 & u_old,
+                                               const T4 & u_older) const
+{
+  // computing first derivative
+  // using the Central Difference method
+  // u_dot_old = (first_term - second_term) / 2 / dt
+  //       first_term = u
+  //      second_term = u_older
+  u_dot -= u_older; // 'older than older' solution
+  u_dot *= 1.0 / (2.0 * _dt);
+
+  // computing second derivative
+  // using the Central Difference method
+  // u_dotdot_old = (first_term - second_term + third_term) / dt / dt
+  //       first_term = u
+  //      second_term = 2 * u_old
+  //       third_term = u_older
+  u_dotdot -= u_old;
+  u_dotdot -= u_old;
+  u_dotdot += u_older;
+  u_dotdot *= 1.0 / (_dt * _dt);
+}
```
How does this even work not in a header file?

ddm42 replied:
Why would implementation need to be in the header file? I moved it because that seemed wrong to me. But I can move it back if there was a reason.

hugary1995 replied:
See this answer: https://stackoverflow.com/a/495056

ddm42 replied:
I see. Then yes I'm surprised it still works. I'll move it back in my next push.

hugary1995 replied:
well it works probably because people never instantiate the template types to be anything other than NumericVector (or whatever that is). If that can be guaranteed, you can remove the template and separate the declaration and implementation. Otherwise, I suggest putting it back to the header.

lynnmunday commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1082,8 +1084,9 @@ FEProblemBase::initialSetup()
 
   if (!_app.isRecovering() && !_app.isRestarting())
```
```suggestion
  if (!_app.isRecovering() && !_app.isRestarting())
```

lynnmunday replied:
```suggestion
  if (!_app.isRecovering() && !_app.isRestarting())
```

lynnmunday replied:
delete if statement

lynnmunday commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1082,8 +1084,9 @@ FEProblemBase::initialSetup()
+    // This functionality was moved to FEProblemBase.C::projectSolution()
+    // // During initial setup the solution is copied to the older solution states (old, older, etc)
+    // // copySolutionsBackwards();
   }
```
```suggestion
  
```

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3332,64 +3328,86 @@ FEProblemBase::projectSolution()
 
-  _aux->solution().close();
-  _aux->solution().localize(*_aux->sys().current_local_solution, _aux->dofMap().get_send_list());
+  // Resetting _global_current_state to 0 in case anyone needs to run ICs again (which they do
+  // sometimes).
```
In the comment, can you spell out the situations in which people want to run ICs again?

hugary1995 commented on framework/src/variables/MooseVariableData.C:
```diff
@@ -9,6 +9,7 @@
 
 #include "MooseVariableData.h"
 #include "MooseVariableField.h"
+#include "MooseVariableDataBase.h"
```
```suggestion
```

ddm42 commented on framework/src/loops/ComputeInitialConditionThread.C:
```diff
@@ -93,8 +93,23 @@ ComputeInitialConditionThread::operator()(const ConstElemRange & range)
 
-      for (auto ic : vec)
-        ic->compute();
+      /* Logic for running only old state initial conditions if they exist, but
+          running ICs normally if the old state ICs don't exist.
+      */
+      bool break_loop = false;
+      int loop_state;
+      for (loop_state = _fe_problem.getGlobalCurrentState(); loop_state >= 0; loop_state--)
+      {
+        for (auto ic : vec)
+          if (ic->getMyState() == loop_state)
+          {
+            ic->compute();
+            break_loop = true;
+          }
+
+        if (break_loop == true)
+          break;
+      }
```
It is called once for each state.

ddm42 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -3332,64 +3328,86 @@ FEProblemBase::projectSolution()
-  if (haveFV())
+  // loop over _global_current_state = 2, 1, 0
+  for (_global_current_state = 2; _global_current_state >= 0; _global_current_state--)
   {
-    using ElemInfoRange = StoredRange<MooseMesh::const_elem_info_iterator, const ElemInfo *>;
-    ElemInfoRange elem_info_range(_mesh.ownedElemInfoBegin(), _mesh.ownedElemInfoEnd());
+    ComputeInitialConditionThread cic(*this);
+    Threads::parallel_reduce(elem_range, cic);
```
Yes. This was required to avoid more complicated logic that would be prone to more edge cases. Here are the design requirements:

1. _global_current_state has to exist inside FEProblemBase. This was to respect the ordering of IC computing that happens in FEProblemBase::projectSolution(). In that function, someone decided that there was a particular order that ICs needed to be computed in. However, inside ComputeInitialConditionThread::compute() was the approperiate place to apply the logic of which ICs to compute. 

2. And then, sys->copySolutionsBackwards() needs to be called from FEProblemBase, as far as I can tell. 

This led me to need some way of transferring loop_state information back and forth between FEProblemBase::projectSolution() and ComputeInitialConditionThread::compute().

3. The final requirement was that if someone only specifies initial conditions for the current state, that IC needs to be copied backward to the old and older states. This was the old functionality of MOOSE.
4. Otherwise, if someone specified multiple states, then the proper ICs needed to be copied back to their appropriate states. 

So if there is a better way to accomplish these things, I'm open to suggestions.

The way the loop inside ComputeInitialConditionThread::compute() works is that it checks to see if state=2 was specified for the variable, then state=1, then state=0. If state=2 was specified, then it will abort the loop, and the state=2 initial condition will get copied back to the appropriate state. But if someone only specified state=0 for that state, then the loop won't find any initial conditions for states 1 and 2, so it will choose the state=0 IC, and that will get copied back to the old states.

hugary1995 replied:
1 & 2: sure, I didn't complain about the way you transfer state between FEProblem and ICs...

3 & 4 should definitely be documented somewhere. I personally prefer that to be part of the IC system documentation.

> The way the loop inside ComputeInitialConditionThread::compute() works is that it checks to see if state=2 was specified for the variable, then state=1, then state=0. If state=2 was specified, then it will abort the loop, and the state=2 initial condition will get copied back to the appropriate state. But if someone only specified state=0 for that state, then the loop won't find any initial conditions for states 1 and 2, so it will choose the state=0 IC, and that will get copied back to the old states.

I think I understand what you want to achieve now. I'd like to suggest a slightly different implementation because your current implementation computes the IC three times if only state=0 is set for a variable, once for state=0, once for state=1, and once for state=2. If the IC computation is very expensive, you have to evaluate the expensive IC three times!

How about this? Let's first compute IC like we used to, that is, compute IC at state=0, copy the solution vector backwards to the old solution vector (**not the older!**). Then, go to state=1, compute ICs with state=1, and copy the old solution vector to the older solution vector. Finally, go to state=2, compute ICs with state=2. This way you still preserve the old behavior when only state=0 is set for an IC, and you can avoid excessive IC evaluations by simply copying the solution vectors backwards.

The challenge with this new approach is that you need to let the threaded IC loop directly write into old and older solution vectors, instead of what you are currently doing (i.e., always writing to the current solution vector, and copy it backwards).

ddm42 replied:
Great! Actually that would be better, but in my first attempt at implementing this, I was unable to the write the old and older state solution vectors directly. Because they seem to live inside LibMesh or PetsC. Do you have any templates that I could look at without breaking a lot of Moose things? Last time there seemed to be no way to make a new vector before the old state vector was created.

hugary1995 replied:
I'm not terribly familiar with that part. Could you perhaps create the old and older vectors earlier?

@dschwen any pointers?

ddm42 replied:
I worked with him on this, and he seemed to think copying backwards seemed to be the best method. But yes @dschwen please weigh in when you get back from vacation Thursday. I'll attempt to implement this before then. Thank you for the advice @hugary1995

ddm42 replied:
@dschwen and I looked at it together, and there are several layers of APIs to go through to accomplish the "writing to old solution states." I will instead implement some logic to only compute each IC object once.

ddm42 commented on framework/include/ics/InitialConditionBase.h:
```diff
@@ -91,13 +91,18 @@ class InitialConditionBase : public MooseObject,
 
   virtual const std::set<std::string> & getSuppliedItems() override;
 
+  int & getMyState();
```
I tried this, but when looping through the states, it leaves more room for errors. When the unsigned int is reduced each time during the for loops, eventually it wants to become negative, indicating the end of the loop has been reached. But because it is unsigned, the remains positive and causes an endless loop. There are solutions for this of course, but it will create room for mistakes if someone wants to change the logic. I think we should leave all of them as signed ints.

ddm42 replied:
I'll implement it this way. But I respectfully disagree. Casting to signed ints explicitly just for the comparison operators, or adding 1 and ending the loop at 1 are the two simple ways of fixing the problem. And both seem less perfect than having signed ints. Especially when someone might make the mistake again in the future. We could even save more storage space by making them shorts or bytes in if the data is what you're concerned about.

ddm42 replied:
Hopefully my concerns won't be important anymore though when I implement your new logic for applying old and older ICs.

ddm42 commented on modules/solid_mechanics/test/tests/old_state_ic/velocity_ic.i:
```diff
@@ -0,0 +1,143 @@
+  []
+[]
+
+[ICs]
+  [current]
+    type = ConstantIC
+    variable = disp_x
+    value = 0
+    state = CURRENT
+  []
+  [old]
+    type = ConstantIC
+    variable = disp_x
+    value = -1
+    state = OLD
+  []
+[]
```
Alright I added some tests to check OLD and OLDER. Required editing CopyValueAux. And I added some tests to make sure it's working properly.

ddm42 commented:
> I just had some minor suggestions. Though we should probably double (triple) check the loop logic as I pointed out in the comments. I feel like your loop is computing certain variable ICs some extra times. Even though that doesn't matter from the correctness standpoint, that's still some extra (wasted) computation...
> 
> Other than that just add one more test for `OLDER` and I'll approve this.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2401628/) on a4b318f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28083/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28083/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 31532df1bd0265aef00ad80d6c0b7639bfee5a13`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2405344/) on f95c07b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>78f728</th>
      <th colspan=4>#28083 f95c07</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.01%</td>
      <td>85.00%</td>
      <td>-0.01%</td>
      <td>68.83%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>105092</td>
      <td>105138</td>
      <td>+46</td>
      <td>53</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18528</td>
      <td>18550</td>
      <td>+22</td>
      <td>24</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28083/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28083/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 68.83% is less than the suggested 90.0%



This comment will be updated on new commits.

dschwen commented:
We probably need to move a test from modules into framework.

dschwen commented on modules/solid_mechanics/test/tests/old_state_ic/tests:
```diff
@@ -0,0 +1,27 @@
+  [old_state_ic]
+    type = CSVDiff
+    input = 'old_state_ic.i'
+    csvdiff = 'old_state_ic_out.csv'
```
Missing `design` documentation link.

maxnezdyur commented on framework/include/auxkernels/CopyValueAux.h:
```diff
@@ -24,8 +24,13 @@ class CopyValueAux : public AuxKernel
+  unsigned short _state;
+
   /// The variable to project from
   const VariableValue & _v;
```
Get rid of all the other VariableValues now. With the change I sent in the .C file you don't need to store all of these anymore.

maxnezdyur commented on framework/src/auxkernels/CopyValueAux.C:
```diff
@@ -19,11 +19,22 @@ CopyValueAux::validParams()
-  : AuxKernel(parameters), _v(coupledValue("source")), _source_variable(*getVar("source", 0))
+  : AuxKernel(parameters),
+    _state(getParam<MooseEnum>("state")),
+    _v(coupledValue("source")),
+    _v_old(_state == 1 ? coupledValueOld("source") : _v),
+    _v_older(_state == 2 ? coupledValueOlder("source") : _v),
```
```suggestion
 _v(_state == 0   ? coupledValue("source")
       : _state == 1 ? coupledValueOld("source")
                     : coupledValueOlder("source")),
```
If you use the above then you do not need to chang the computeValue method that was originally there.

maxnezdyur reviewed (changes requested):
In `framework/doc/content/syntax/ICs/index.md`

add something like 
```
## Old and Older ICs

The ICs system supports the ability to set ICs on old and older states. This can be useful for setting initial velocity or acceleration conditions. It can be set with the [!param](/ICs/ConstantIC/state) parameter and specifying `OLD` or `OLDER`.
```
to tell users about the new feature and how to use it.

maxnezdyur committed:
fix docs and simplify copyAux

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2625976/), step Generate coverage on ae8a56a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>1e1768</th>
      <th colspan=4>#28083 ae8a56</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.25%</td>
      <td>85.24%</td>
      <td>-0.01%</td>
      <td>67.14%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108011</td>
      <td>108050</td>
      <td>+39</td>
      <td>47</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18681</td>
      <td>18704</td>
      <td>+23</td>
      <td>23</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28083/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28083/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28083/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 67.14% is less than the suggested 90.0%



This comment will be updated on new commits.

hugary1995 commented on framework/doc/content/syntax/ICs/index.md:
```diff
@@ -62,6 +62,13 @@ object's variable *must* be defined on the block - otherwise the IC for the
 
+## Old and Older ICs
+
+The ICs system supports the ability to set ICs on old and older states. This can be useful for setting initial velocity or acceleration conditions. It can be set with the [!param](/ICs/ConstantIC/state) parameter and specifying `OLD` or `OLDER`.
```
```suggestion
The ICs system supports the ability to set ICs on old and older states. This can be useful for initializing old and older states of variables needed for various time integration schemes. It can be set with the [!param](/ICs/ConstantIC/state) parameter and specifying `OLD` or `OLDER`.
```

hugary1995 commented on framework/include/ics/InitialConditionInterface.h:
```diff
@@ -0,0 +1,37 @@
+
+  static InputParameters validParams();
+
+  unsigned short & getState();
```
Add a doxygen comment, especially about the fact that the returned reference is writable.

hugary1995 commented on framework/include/ics/InitialConditionWarehouse.h:
```diff
@@ -40,8 +40,13 @@ class InitialConditionWarehouse : public MooseObjectWarehouseBase<InitialConditi
+  /// Each map relates string-int tuples to a set of block/boundary IDs.
+  /// The string-int tuple is a unique identifier for a specific variable and state. The string-int tuple is renamed to ic_key_type for clarity.
+  /// The algorithm then makes sure that a new IC object does not overlap with a previous IC object (i.e. same block/boundary).
+  typedef std::tuple<std::string, unsigned short> ic_key_type;
```
I'd prefer `using` over `typedef`, so something like
```cpp
using ic_key_type = std::tuple<VariableName, unsigned short>;
```

hugary1995 commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -884,6 +884,8 @@ class FEProblemBase : public SubProblem, public Restartable
 
   void projectSolution();
 
+  unsigned short getCurrentICState();
```
Add a doxygen comment.

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1084,6 +1085,65 @@ FEProblemBase::initialSetup()
+      // save state 0
+      for (const auto i : index_range(_solver_systems))
+      {
+        _solver_systems[i]->system().update(); // not sure if this update is needed
```
I don't think this is needed. Same applies to the following system updates. If any test fails without this call, I'd like to understand why.

hugary1995 commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -1084,6 +1085,65 @@ FEProblemBase::initialSetup()
+      _aux->system().update(); // not sure if this update is needed
+      state0_aux_buffer = _aux->solutionState(0).clone();
+
+      // compute old state ICs
+      for (_current_ic_state = 1; _current_ic_state <= ic_state_max; _current_ic_state++)
+      {
+        projectSolution();
+
+        for (auto & sys : _solver_systems)
+        {
+          sys->system().update(); // not sure if this update is needed
+          sys->solutionState(_current_ic_state) = sys->solutionState(0);
+        }
+        _aux->system().update(); // not sure if this update is needed
+        _aux->solutionState(_current_ic_state) = _aux->solutionState(0);
+      }
+      _current_ic_state = 0;
+
+      // recover state 0
+      for (const auto i : index_range(_solver_systems))
+      {
+        _solver_systems[i]->solutionState(0) = *state0_sys_buffers[i];
+        _solver_systems[i]->solutionState(0).close();
+        _solver_systems[i]->update();
+      }
+      _aux->solutionState(0) = *state0_aux_buffer;
+      _aux->solutionState(0).close();
+      _aux->update();
```
This logic now looks correct to me.

hugary1995 reviewed (changes requested):
I have some minor comments. Did you also add a test for state=OLDER?

maxnezdyur commented:
> I have some minor comments. Did you also add a test for state=OLDER?

The input file`old_state_ic.i` applies an different IC value for all 3 states.

maxnezdyur committed:
address review comments, clean up updates
